src/plugins/english/divinedaolibrary.ts

```diff
@@ -1,125 +1,259 @@
+        if (!path) return;
+        return {
+          name: loadedCheerio(anchorEl).text(),
+          path,
+        };
+      })
+      .toArray()
+      .filter(
+        (novel, index, all) =>
+          all.findIndex(({ path }) => path === novel.path) === index,
+      );
+    return novels;
   }
 
-  async parseNovel(novelPath: string): Promise<Plugin.SourceNovel> {
-    const result = await fetchApi(this.site + novelPath);
-    const body = await result.text();
-
+  /**
+   * Parse list of names, paths, and categories from the novels list.
+   *
+   * @private
+   */
+  async allNovels() {
+    const body = await fetchApi(this.site + 'novels').then(res => res.text());
     const loadedCheerio = parseHTML(body);
+    const novels = loadedCheerio('.entry-content ul')
+      .map((_, listEl) => {
+        const list = loadedCheerio(listEl);
+        const category = list.prev().text();
+        return list
+          .find('a')
+          .map((_, anchorEl) => {
+            const path = this.getPath(anchorEl.attribs['href']);
+            if (!path) return;
+            return {
+              name: loadedCheerio(anchorEl).text(),
+              path,
+              category,
+            };
+          })
+          .toArray();
+      })
+      .toArray();
+    return novels;
+  }
 
-    const novel: Plugin.SourceNovel = {
-      path: novelPath,
-      name: loadedCheerio('h1.entry-title').text().trim() || 'Untitled',
-      cover:
-        loadedCheerio('.entry-content').find('img').attr('data-ezsrc') ||
-        defaultCover,
-      chapters: [],
-    };
-
-    novel.summary = loadedCheerio('#main > article > div > p:nth-child(6)')
-      .text()
-      .trim();
-
-    novel.author = loadedCheerio('#main > article > div > h3:nth-child(2)')
-      .text()
-      .replace(/Author:/g, '')
-      .trim();
-
-    const chapter: Plugin.ChapterItem[] = [];
-
-    loadedCheerio('#main')
-      .find('li > span > a')
-      .each((i, el) => {
-        const chapterName = loadedCheerio(el).text().trim();
-        const chapterUrl = loadedCheerio(el).attr('href');
-
-        if (!chapterUrl) return;
-
-        chapter.push({
-          name: chapterName,
-          path: chapterUrl.replace(this.site, ''),
-        });
-      });
-
-    novel.chapters = chapter;
+  async popularNovels(
+    pageNo: number,
+    options: Plugin.PopularNovelsOptions<Filters>,
+  ): Promise<Plugin.NovelItem[]> {
+    if (pageNo !== 1) {
+      return [];
+    }
+    if (options.showLatestNovels) {
+      return await this.asyncMap(
+        await this.latestNovels(),
+        this.grabNovel.bind(this),
+      );
+    }
+    const novels = (await this.allNovels()).filter(novel => {
+      const selectedCategories = options.filters.category.value;
+      if (Array.isArray(selectedCategories)) {
+        return selectedCategories.includes(novel.category);
+      }
+      return false;
+    });
+    return await this.asyncMap(novels, this.grabNovel.bind(this));
```

Comment:
I see that you fetch detail for every novel in popular / search to get covers, which is quite expensive. 
I think you could create a temp cache by adding a `Set` in plugin instance for caching novel detail.

Reply:
this would save lot of time if user visit DDL multiple times

Reply:
Happy to add some caching. Is there a best practice that plugins usually follow? Would you cache the full parsed page (including author and summary) or only the cover URL for display on the novel list?

I felt the few extra requests were worth it to me to make it look better, especially since there are relatively few novels and I thought it unlikely anyone was going to visit it often.

Reply:
You can just store it in a variable (memory) which would be lost when user turn off the app. 

Or you can use plugin storage ([example](https://github.com/LNReader/lnreader-plugins/blob/master/src/plugins/russian/authortoday.ts#L7)). But dont abuse it, storing large amout of data could hurt app performance.

Btw app just need novel name, path, and cover to display novels  in  browse and search screen. So I think you just need to cache cover urls

Reply:
Is storage kept between app starts? That does sound like overkill for what is really just an optimisation for the number of requests. I would expect people to rarely go back into the novel list. Optimising for people who are looking to select a novel makes a lot more sense.

Not sure I really like just randomly stuffing information in memory that does not get cleared until application restart either. But I guess the application is keeping an instance of every installed plugin in memory anyway? So it really wouldn’t be that hard.

I’ll have a go at adding a cache for reading the novel pages later today and get this PR updated.

Reply:
Is storage kept between app starts? That does sound like overkill for what is really just an optimisation for the number of requests. I would expect people to rarely go back into the novel list. Optimising for people who are looking to select a novel makes a lot more sense.

Not sure I really like just randomly stuffing information in memory that does not get cleared until application restart either. But I guess the application is keeping an instance of every installed plugin in memory anyway? So it really wouldn’t be that hard.

I’ll have a go at adding a cache for reading the novel pages later today and get this PR updated.